scenario_player/main.py

```diff
@@ -220,94 +224,158 @@ def run(
+        runner_manager = ScenarioRunnerManager(scenario_runner)
+        notify_setting = runner_manager.scenario_runner.definition.settings.notify
+        reporter = report_result(report_container, notify_setting, mailgun_api_key)
+        with reporter, runner_manager as runner, service_manager:
+            log.info("Startup complete")
+            if enable_ui:
+                log_buffer = attach_urwid_logbuffer()
+                ui = ScenarioUIManager(runner, log_buffer, log_file_name, success)
+            else:
+                ui = nullcontext()
+            with ui:
+                runner.run_scenario()
+    except Exception:
+        raise
+
+
+class ScenarioUIManager:
+    def __init__(self, runner, log_buffer, log_file_name, success):
+        self.ui = ScenarioUI(runner, log_buffer, log_file_name)
+        self.success = success
+        self.ui_greenlet = self.ui.run()
+
+    def __enter__(self):
+        return self.success
+
+    def __exit__(self, type, value, traceback):
         try:
-            if enable_ui and ui:
-                ui.set_success(success)
-                log.warning("Press q to exit")
-                while not ui_greenlet.dead:
-                    gevent.sleep(1)
-            service_process.stop()
-        except ServiceProcessException:
-            service_process.kill()
+            self.ui.set_success(self.success.get())
+            log.warning("Press q to exit")
+            while not self.ui_greenlet.dead:
+                gevent.sleep(1)
         finally:
-            runner.node_controller.stop()
-            if ui_greenlet is not None and not ui_greenlet.dead:
-                ui_greenlet.kill(ExitMainLoop)
-                ui_greenlet.join()
-            exit(exit_code)
+            if self.ui_greenlet is not None and not self.ui_greenlet.dead:
+                self.ui_greenlet.kill(ExitMainLoop)
+                self.ui_greenlet.join()
+
+
+class ServiceProcessManager:
+    def __init__(self):
+        self.service_process = ServiceProcess()
+        self.service_process.start()
+
+    def __enter__(self):
+        return self.service_process
+
+    def __exit__(self, type, value, traceback):
+        try:
+            self.service_process.stop()
+        except ServiceProcessException:
+            log.exception("ServiceProcessManager died")
+        except Exception:
+            self.service_process.kill()
+
+
+class ScenarioRunnerManager:
+    def __init__(self, scenario_runner: ScenarioRunner):
+        self.scenario_runner = scenario_runner
+
+    def __enter__(self):
+        return self.scenario_runner
+
+    def __exit__(self, type, value, traceback):
+        try:
+            self.scenario_runner.node_controller.stop()
+        except ProcessExitedWithError:
+            log.exception("ScenarioRunnerManager close died")
+        except Exception:
+            self.scenario_runner.node_controller.kill()
+
+
+@contextmanager
+def report_result(container, runner, mailgun_api_key):
```

Comment:
Any reason for using `@contextmanager` here but not in the two context managers above?

Reply:
this is, as far as I understand, the "second style" of a context manager:
1) class with `__enter__` & `__exit__`
2) `yield`ing function with the decorator

Reply:
With the current code, you could `@contextmanager ` for all context managers. That would be more consistent and a bit more concise.